src/mutable_data.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright 2019 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::XorName;
+use std::collections::{BTreeMap, BTreeSet};
+use std::vec::Vec;
+// use threshold_crypto::PublicKey;
+use rust_sodium::crypto::sign::PublicKey;
+
+/// Used to differentiate between conflict-free and conflict aware variants.
+#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, Serialize, Deserialize)]
+pub enum MutableDataKind {
+    // Unsequenced, unpublished Mutable Data.
+    Unsequenced { data: BTreeMap<Vec<u8>, Value> },
+    // Sequenced, unpublished Mutable Data.
+    Sequenced { data: BTreeMap<Vec<u8>, Vec<u8>> },
+}
+
+/// Permissions given to a user / application for a mutable data field.
+#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, Serialize, Deserialize)]
+pub enum Permission {
+    Read,
+    Insert,
+    Update,
+    Delete,
+    ManagePermissions,
+}
+
+/// Mutable data that is unpublished on the network. This data can only be fetch be the owners / those in the permissions fiedls with `Permission::Read` access.
+#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, Serialize, Deserialize)]
+pub struct UnpublishedMutableData {
```

Comment:
I'd say let's just leave it at just `MutableData` since it is not a application developer facing API. It is implied that MutableData is unpublished, so this will be an unnecessary qualification here.

Reply:
I left the prefix in there to differentiate this from the existing data type in routing. But I guess we can always rename the import to show the difference :+1:

Reply:
I also proposed this change to the RFC in case you guys haven't seen it:
https://github.com/maidsafe/rfcs/pull/309